compiler/src/codegen/compcore.re

```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
+  // contributor document.
+
+  // Extension constructors defined by the module must be grouped together.
+  // For now, this only includes exceptions.
+  let (exception_meta, non_exception_meta) =
+    List.partition(
+      fun
+      | ExceptionMetadata(_) => true
+      | _ => false,
+      type_metadata,
+    );
+  let type_metadata =
+    switch (exception_meta) {
+    | [ExceptionMetadata(id, _, _), ..._] => [
+        ADTMetadata(
+          id,
+          List.map(
+            meta => {
+              switch (meta) {
+              | ExceptionMetadata(_, variant, name) => (variant, name)
+              | _ => failwith("impossible by partition")
+              }
+            },
+            exception_meta,
+          ),
+        ),
+        ...non_exception_meta,
+      ]
+    | _ => non_exception_meta
+    };
+
+  let round_to_8 = n => Int.logand(n + 7, Int.lognot(7));
+
+  let buf = Buffer.create(256);
+  Buffer.add_int32_le(buf, 0l); // Slot where pointer to next will live
+  Buffer.add_int32_le(buf, 0l); // Slot where module id will live
+  Buffer.add_int32_le(buf, Int32.of_int(List.length(type_metadata)));
+
+  List.iter(
+    meta => {
+      switch (meta) {
+      | ADTMetadata(id, cstrs) =>
+        let section_length =
+          List.fold_left(
+            (total, (_, cstr)) =>
+              total + 12 + round_to_8(String.length(cstr)),
+            8,
+            cstrs,
+          );
+        Buffer.add_int32_le(buf, Int32.of_int(section_length));
+        Buffer.add_int32_le(buf, Int32.of_int(id));
+        List.iter(
+          ((id, cstr)) => {
+            let length = String.length(cstr);
+            let aligned_length = round_to_8(length);
+            Buffer.add_int32_le(buf, Int32.of_int(aligned_length + 12));
+            Buffer.add_int32_le(buf, Int32.of_int(id));
+            Buffer.add_int32_le(buf, Int32.of_int(length));
+            Buffer.add_string(buf, cstr);
+            for (_ in 1 to aligned_length - length) {
+              Buffer.add_int8(buf, 0);
+            };
+          },
+          cstrs,
+        );
+      | RecordMetadata(id, fields) =>
+        let section_length =
+          List.fold_left(
+            (total, field) => total + 8 + round_to_8(String.length(field)),
+            8,
+            fields,
+          );
+        Buffer.add_int32_le(buf, Int32.of_int(section_length));
+        Buffer.add_int32_le(buf, Int32.of_int(id));
+        List.iter(
+          field => {
+            let length = String.length(field);
+            let aligned_length = round_to_8(length);
+            Buffer.add_int32_le(buf, Int32.of_int(aligned_length + 8));
+            Buffer.add_int32_le(buf, Int32.of_int(length));
+            Buffer.add_string(buf, field);
+            for (_ in 1 to aligned_length - length) {
+              Buffer.add_int8(buf, 0);
+            };
+          },
+          fields,
+        );
+      | ExceptionMetadata(_) => failwith("impossible by partition")
```

Comment:
🤔  When I wrote this I thought the opposite, that I shouldn't bother with making `ExceptionMetadata` since it would just largely be the same as `ADTMetadata`. The partition I'm talking about is the one that happens right above this block of code, and that's just done because there's some preprocessing that needs to happen to them anyway, and after that it's the same as `ADTMetadata`. We could instead do the preprocessing in `linearize.re`, but you'd still need some sort of type to collect the values so you could process them. It also felt like the distinction was better made here, since it's really just that at the code level they're the same, but different elsewhere. I dunno.

Reply:
Yeah, I understand, but I've found my conceptualization of ADTs to be heavily influenced by the talk "Making Impossible States Impossible" by Richard Feldman

It just feels weird that we have this ADT that we massage and then put impossible cases in our pattern matches.

Reply:
Yeah. FWIW, the reason why we do that a lot in other places is because we create some types in the parsetree, and then every other stage of the compiler we alias the type instead of creating a new one with just the variants we need. We should probably make an issue to reconsider that.

I can sort of do what you're saying here to get rid of this case.

Reply:
Up to you, I know we do it a lot and it doesn't actually affect the correctness of the compiler. I would be happy with an issue!